compiler/rustc_typeck/src/collect.rs

```diff
@@ -2469,43 +2469,187 @@ fn compute_sig_of_foreign_fn_decl<'tcx>(
         None,
     );
 
-    // Feature gate SIMD types in FFI, since I am not sure that the
-    // ABIs are handled at all correctly. -huonw
-    if abi != abi::Abi::RustIntrinsic
-        && abi != abi::Abi::PlatformIntrinsic
-        && !tcx.features().simd_ffi
-    {
-        let check = |ast_ty: &hir::Ty<'_>, ty: Ty<'_>| {
-            if ty.is_simd() {
-                let snip = tcx
-                    .sess
-                    .source_map()
-                    .span_to_snippet(ast_ty.span)
-                    .map_or_else(|_| String::new(), |s| format!(" `{}`", s));
-                tcx.sess
-                    .struct_span_err(
-                        ast_ty.span,
-                        &format!(
-                            "use of SIMD type{} in FFI is highly experimental and \
-                             may result in invalid code",
-                            snip
-                        ),
-                    )
-                    .help("add `#![feature(simd_ffi)]` to the crate attributes to enable")
-                    .emit();
-            }
-        };
+    // Using SIMD types in FFI signatures requires the signature
+    // to have appropriate `#[target_feature]`'s enabled.
+    if abi != abi::Abi::RustIntrinsic && abi != abi::Abi::PlatformIntrinsic {
         for (input, ty) in iter::zip(decl.inputs, fty.inputs().skip_binder()) {
-            check(&input, ty)
+            simd_ffi_check(tcx, def_id, &input, ty)
         }
         if let hir::FnRetTy::Return(ref ty) = decl.output {
-            check(&ty, fty.output().skip_binder())
+            simd_ffi_check(tcx, def_id, &ty, fty.output().skip_binder())
         }
     }
 
     fty
 }
 
+/// Returns `Ok()` if the target-feature allows using the SIMD type on C FFI.
+/// Otherwise, returns `Err(Some())` if the target_feature needs to be enabled or
+/// or `Err(None)` if it's unsupported.
+fn simd_ffi_feature_check(
```

Comment:
I don't know enough about simd to properly review this function. @lqd can you help or ping others?

That said, the function body looks very fragile and easy to get accidentally wrong to me. Is there something declarative that we could parse and process here?

Reply:
I can help ... by asking the 2 greats @Amanieu and @nagisa for help ^^

Reply:
> That said, the function body looks very fragile and easy to get accidentally wrong to me. Is there something declarative that we could parse and process here?

I agree but I haven't come up with/know a better solution. @Amanieu Do you have any ideas?

Reply:
The true source of these ABI constraints is in LLVM's handling of calling conventions. I don't think there is anything we can easily parse here.

My recommendation would be to directly refer to LLVM's source code for calling convention handling to determine which target features affect ABI.

In a way this is similar to what I had to do for `asm!` in `compiler/rustc_target/src/asm` which is similarly tied to LLVM implementation details for reserved registers.

Reply:
Couldn't LLVM be extended to provide APIs to query information about calling conventions instead of relying on its source code, which is also fragile.